site/docs/plugins/runner.md

```diff
@@ -78,10 +80,299 @@ run(bot);
 </CodeGroupItem>
 </CodeGroup>
 
-Of course, while this looks very simple, a lot is going on under the hood.
+## Sequential Processing Where Necessary
+
+Most likely, you want to be guaranteed that messages from the same chat are processed in order.
+This is useful when installing [session middleware](./session.md), but it also makes sure that your bot does not confuse the order of messages in the same chat.
+
+grammY runner exports the `sequentialize` middleware that takes care of this.
+You can check out this [section](../advanced/scaling.md#concurrency-is-hard) to learn how to use it.
+
+We are now going to look at more advanced usage of the plugin.
+
+The supplied constraint function can be used not only to specify chat identifier, or user identifier.
+Instead, you can return _a list of constraint identifier strings_ that determine for every update individually what other computations it must wait for before processing can begin.
+
+As an example, you could return both the chat identifier, and the user identifier of the message author.
+
+```ts
+bot.use(sequentialize((ctx) => {
+  const chat = ctx.chat?.id.toString();
+  const user = ctx.from?.id.toString();
+  return [chat, user].filter((con) => con !== undefined);
+}));
+```
+
+This would make sure that messages in the same chat are ordered correctly.
+In addition, if Alice sends message in a group, and then sends a message to your bot in the private chat, then these two messages are ordered correctly.
+
+In a sense, you can therefore specify a graph of dependencies between updates.
+grammY runner will resolve all necessary constraints on the fly and block those updates as long as necessary to ensure correct message ordering.
+
+The implementation of this is very efficient.
+It needs constant memory (unless you specify infinite concurrency), and it needs (amortized) constant processing time per update.
+
+## Graceful Shutdown
+
+In order for the bot to complete its work correctly you [should signal](../advanced/reliability.md#using-grammy-runner) it to stop when the process is about to be destroyed.
+
+Note that you can wait for the runner to terminate by `await`ing the `task` in the [`RunnerHandle`](https://deno.land/x/grammy_runner/mod.ts?s=RunnerHandle) returned from `run`.
+
+```ts
+const handle = run(bot);
+
+handle.task.then(() => {
+  console.log("Bot done processing!");
+});
+```
+
+## Advanced Options
+
+grammY runner consists of three things: a source, a runner, and a sink.
```

Comment:
Maybe it’ll be cooler like this since it will match the order of the explanation following it?

```suggestion
grammY runner consists of three things: a source, a sink, and a sink.
```

Reply:
It will have to be this:

```suggestion
grammY runner consists of three things: a source, a sink, and a runner.
```